### PR TITLE
Added library version

### PR DESCRIPTION
**Pull request by theapache64**
With this developers can convert `JSON` to `Kotlin` data class programmatically.

- Pending Task : Upload to jcenter/maven

Ref : #258

### CODE REVIEW

File: library/src/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+     * Default : DefaultValueStrategy.AvoidNull
+     */
+    fun setDefaultValueStrategy(strategy: DefaultValueStrategy): JsonToKotlinBuilder {
+        TestConfig.defaultValueStrategy = strategy
+        return this
+    }
+
+    /**
+     * To set JSON decoding/encoding library
+     * Default: TargetJsonConverter.Gson
```
Comment (wuseal) [inline library/src/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt]: I think the default value should be `TargetJsonConverter.None`, Keep the same with plugin

File: library/src/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+            isTestModel = true
+            isCommentOff = false
+            isOrderByAlphabetical = true
+            isPropertiesVar = false
+            targetJsonConvertLib = TargetJsonConverter.NoneWithCamelCase
+            propertyTypeStrategy = PropertyTypeStrategy.AutoDeterMineNullableOrNot
+            defaultValueStrategy = DefaultValueStrategy.None
+            isNestedClassModel = true
+            customPropertyAnnotationFormatString = "@Optional\n@SerialName(\"%s\")"
```
Comment (wuseal) [inline library/src/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt]: If you don't choose `TargetJsonConverter.Custom`, There is no need to give value to customPropertyAnnotation*

Comment (theapache64) [reply library/src/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt]: I just copied it from the `TestConfig.kt` 

https://github.com/wuseal/JsonToKotlinClass/blob/51245c785208e421dd199c6fb8141ca9510f6135/src/main/kotlin/wu/seal/jsontokotlin/test/TestConfig.kt#L29

Comment (wuseal) [reply library/src/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt]: @theapache64 Got it, Although there is no problem did as this

Comment (wuseal) [reply library/src/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt]: Yes! Remove it.

File: library/src/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt (hunk 1 above)
Comment (theapache64) [reply library/src/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt]: It's `None'. Need to change just the comment. Will do it.

File: library/src/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt (hunk 2 above)
Comment (wuseal) [reply library/src/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt]: Also, where are the methods about setting these properties: `customPropertyAnnotationFormatString` `customAnnotaionImportClassString` `customClassAnnotationFormatString`. they are need when set TargetJsonConverter to `Custom`

Comment (theapache64) [reply library/src/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt]: Ok. Will add

File: library/src/main/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+    init {
+        TestConfig.apply {
+
+            isTestModel = true
+            isCommentOff = true
+            isOrderByAlphabetical = true
+            isPropertiesVar = false
+            targetJsonConvertLib = TargetJsonConverter.None
+            propertyTypeStrategy = PropertyTypeStrategy.AutoDeterMineNullableOrNot
```
Comment (wuseal) [inline library/src/main/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt]: this default should be `PropertyTypeStrategy.NotNull`

File: src/main/kotlin/wu/seal/jsontokotlin/Main.kt [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package wu.seal.jsontokotlin
+
+fun main() {
```
Comment (wuseal) [inline src/main/kotlin/wu/seal/jsontokotlin/Main.kt]: Do we need this file now?

File: LIBRARY.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    .setForcePrimitiveTypeNonNullable(true) // optional, default : false
+    .build(json1, "GlossResponse") // finally, get KotlinClassCode string
+```
+
+## Demo
+
+**Code**
+
+```kotlin
+import com.theapache64.j2k.JsonToKotlinBuilder
```
[the remark this answers is not in the document]
Comment (theapache64) [reply LIBRARY.md]: Yeah. It's updated. Just forgot to change it in the docs.

File: library/src/main/kotlin/wu/seal/jsontokotlin/Test.kt [hunk 6]
```diff
@@ -0,0 +1,51 @@
+import wu.seal.jsontokotlin.JsonToKotlinBuilder
+import wu.seal.jsontokotlin.*
+
+fun main() {
```
[the remark this answers is not in the document]
Comment (theapache64) [reply library/src/main/kotlin/wu/seal/jsontokotlin/Test.kt]: It's written for raw test run. Will delete it. :+1:

File: src/main/kotlin/wu/seal/jsontokotlin/Main.kt (hunk 4 above)
Comment (theapache64) [reply src/main/kotlin/wu/seal/jsontokotlin/Main.kt]: Will delete it.

File: LIBRARY.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    .setComment(true) // optional, default : true
+    .setOrderByAlphabetic(true) // optional : default : true
+    .setInnerClassModel(true) // optional, default : true
+    .setMapType(true)// optional, default : false
+    .setCreateAnnotationOnlyWhenNeeded(true) // optional, default : false
+    .setIndent(4)// optional, default : 4
+    .setParentClassTemplate("android.os.Parcelable") // optional, default : ""
+    .setKeepAnnotationOnClass(true) // optional, default : false
+    .setKeepAnnotationOnClassAndroidX(true) // optional, default : false
+    .setKeepAnnotationAndPropertyInSameLine(true) // optional, default : false
```
[the remark this answers is not in the document]
Comment (theapache64) [reply LIBRARY.md]: So, all methods that receives a single boolean, `setXX(true or false)`, should replace with `enableXXX(true or false)`, right ?

[the remark this answers is not in the document]
Comment (wuseal) [reply LIBRARY.md]: Yes, That's my advice, How do you think?

File: library/src/main/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt (hunk 3 above)
Comment (theapache64) [reply library/src/main/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt]: Okay

File: LIBRARY.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (theapache64) [reply LIBRARY.md]: Awesome

File: library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt [hunk 8]
```diff
@@ -1,12 +1,101 @@
...
+
+            import com.squareup.moshi.Json
+            import com.squareup.moshi.JsonClass
+            import android.os.Parcelable
+            import android.support.annotation.Keep
+            import androidx.annotation.Keep
+            import kotlinx.android.parcel.Parcelize
+            import android.os.Parcelable
+
+            @SuppressLint("ParcelCreator")
```
Comment (wuseal) [inline library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt]: Why there is no Keep Annotation here?

File: library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt [hunk 9]
```diff
@@ -1,12 +1,101 @@
...
+            @Parcelize
+            data class GlossResponseMyClassSuffix(
+                val MyPrefixAuthorsMySuffix: List<AuthorMyClassSuffix> = listOf(),
```
Comment (wuseal) [inline library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt]: There is no Moshi(CodeGen)  Annotation here, Maybe there are some bugs in this program

File: library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt [hunk 10]
```diff
@@ -1,12 +1,101 @@
...
+                .setDefaultValueStrategy(DefaultValueStrategy.AvoidNull) // optional, default : DefaultValueStrategy.AvoidNull
+                .setAnnotationLib(TargetJsonConverter.MoshiCodeGen) // optional, default: TargetJsonConverter.None
+                .enableComments(true) // optional, default : false
+                .enableOrderByAlphabetic(true) // optional : default : false
+                .enableInnerClassModel(true) // optional, default : false
+                .enabelMapType(true)// optional, default : false
+                .enableCreateAnnotationOnlyWhenNeeded(true) // optional, default : false
+                .setIndent(4)// optional, default : 4
+                .setParentClassTemplate("android.os.Parcelable") // optional, default : ""
+                .enableKeepAnnotationOnClass(true) // optional, default : false
```
Comment (wuseal) [inline library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt]: The two keepAnotation should only enable one of them, not all of them, for they are Conflict each other

File: library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt (hunk 9 above)
Comment (wuseal) [reply library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt]: I think this option may lead this bug: `enableCreateAnnotationOnlyWhenNeeded` how about let it stay last.

Comment (wuseal) [reply library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt]: Or we could open a new issue for this problem, then we fix it later.

Comment (theapache64) [reply library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt]: Okay

File: library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt (hunk 10 above)
Comment (theapache64) [reply library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt]: Okay

File: library/src/main/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+
+    fun enableCreateAnnotationOnlyWhenNeeded(isOnlyWhenNeeded: Boolean): JsonToKotlinBuilder {
+        TestConfig.enableMinimalAnnotation = isOnlyWhenNeeded
+        return this
+    }
+
+    /**
+     * To set indent in output kotlin code.
+     *
+     * Default : 4
```
Comment (wuseal) [inline library/src/main/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt]: currently there only this method with test case doc? Do we need to add others?

File: library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt (hunk 8 above)
Comment (theapache64) [reply library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt]: Fixed in #260

File: library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt (hunk 9 above)
Comment (theapache64) [reply library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt]: I think this got fixed in #260 , Isn't it ?

File: library/src/main/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt (hunk 11 above)
Comment (theapache64) [reply library/src/main/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt]: Yes we need to add doc to all methods, but for now, I hope it'll be better if we add docs to methods that are not self explainable. What do you think ?

File: library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt (hunk 9 above)
Comment (wuseal) [reply library/src/test/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilderTest.kt]: yes

File: library/src/main/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt (hunk 11 above)
Comment (wuseal) [reply library/src/main/kotlin/wu/seal/jsontokotlin/JsonToKotlinBuilder.kt]: Yes, U are right

**Pull request merged**